File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/create/CreatePlayListPresenter.java
package com.github.anrimian.musicplayer.ui.playlist_screens.create;

import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;

import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import moxy.MvpPresenter;


public class CreatePlayListPresenter extends MvpPresenter<CreatePlayListView> {

    private final PlayListsInteractor playListsInteractor;
    private final Scheduler uiScheduler;
    private final ErrorParser errorParser;

    private final CompositeDisposable presenterDisposable = new CompositeDisposable();

    public CreatePlayListPresenter(PlayListsInteractor playListsInteractor,
                                   Scheduler uiScheduler,
                                   ErrorParser errorParser) {
        this.playListsInteractor = playListsInteractor;
        this.uiScheduler = uiScheduler;
        this.errorParser = errorParser;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        getViewState().showInputState();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    void onCompleteInputButtonClicked(String playListName) {
        presenterDisposable.add(playListsInteractor.createPlayList(playListName)
                .observeOn(uiScheduler)
                .doOnSubscribe(d -> getViewState().showProgress())
                .subscribe(this::onPlayListCreated, this::onPlayListCreatingError));
    }

    private void onPlayListCreated(PlayList playList) {
        getViewState().onPlayListCreated(playList);
    }

    private void onPlayListCreatingError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showError(errorCommand);
    }
}

File: app/src/main/java/com.github.anrimian.musicplayer/ui/playlist_screens/create/CreatePlayListPresenter.kt
package com.github.anrimian.musicplayer.ui.playlist_screens.create

import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.common.mvp.AppPresenter
import io.reactivex.rxjava3.core.Scheduler

class CreatePlayListPresenter(private val playListsInteractor: PlayListsInteractor,
                              uiScheduler: Scheduler,
                              errorParser: ErrorParser)
    : AppPresenter<CreatePlayListView>(uiScheduler, errorParser) {

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.showInputState()
    }

    fun onCompleteInputButtonClicked(playListName: String) {
        viewState.showProgress()
        playListsInteractor.createPlayList(playListName)
                .subscribeOnUi(this::onPlayListCreated, this::onPlayListCreatingError)
    }

    private fun onPlayListCreated(playList: PlayList) {
        viewState.onPlayListCreated(playList)
    }

    private fun onPlayListCreatingError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showError(errorCommand)
    }

}